I2C default cfg from DT, add support for MCUX from dt

galak opened the pull request.

galak committed:
i2c: mcux: Convert to supporting device tree

Update the MCUX I2C driver and related platforms to get their I2C
information from the device tree.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
i2c: dt: introduce dt properties for setting default cfg

We introduce "zephyr,speed" & "zephyr,use-10-bit-addr" device tree
properties as means to set CONFIG_I2C_n_DEFAULT_CFG.  For now we assume
master mode as the I2C subystem doesn't really support a slave mode.

We update the platforms that use device tree for I2C to get their params
from the DT and setup CONFIG_I2C_n_DEFAULT_CFG.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

ydamigos commented on dts/arm/olimexino_stm32.fixup:
```diff
@@ -31,3 +31,4 @@
 #define CONFIG_I2C_2_NAME			ST_STM32_I2C_V1_40005800_LABEL
 #define CONFIG_I2C_2_EVENT_IRQ			ST_STM32_I2C_V1_40005800_EVENT
 #define CONFIG_I2C_2_ERROR_IRQ			ST_STM32_I2C_V1_40005800_ERROR
+#define CONFIG_I2C_2_DEFAULT_CFG		(ST_STM32_I2C_V2_40005400_ZEPHYR_SPEED << 1) | I2C_MODE_MASTER
```
It should be ST_STM32_I2C_V1_40005800_ZEPHYR_SPEED

galak replied:
Thanks fixed

ydamigos commented:
Tested it on my 96b_carbon. A minor change is needed for olimexino_stm32.

galak committed:
i2c: Fix missing Kconfig !HAS_I2C_DTS dependencies

As I2C_0 (port 0) isn't used on the STM32 platforms we didn't exclude
the related Kconfig options if DTS was enabled.  However other SoCs
(like NXP) do use I2C_0 so we need to fixup the Kconfig dependencies

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

mnkp commented on dts/common/yaml/zephyr-i2c.yaml:
```diff
@@ -0,0 +1,19 @@
+      category: required
+      description: Speed setting (1..5) as defined in i2c.h
+      generation: define
+  - zephyr,use-10-bit-addr :
```
I believe the current support for 10-bit addressing in i2c.h file is incorrect. I2C_ADDR_10_BITS flag is supposed to be used with i2c_configure command and set only once. We treat it the same as speed, master/slave mode. i.e. it's a property which applies to all transactions performed on the bus. That's wrong. From I2C specification:
"Devices with 7-bit and 10-bit addresses can be connected to the same I2C-bus, and both 7-bit and 10-bit addressing can be used in all bus speed modes."
It means that I2C_ADDR_10_BITS should be a property of i2c_msg.flags. In that case every message sent on the I2C bus can have independently set addressing mode.

As such the zephyr,use-10-bit-addr property should be removed from DTS.

erwango replied:
I understand your point, though,here, we're setting a default configuration at boot (as meant initial flag).
This doesn't prevent to change it in run time, as allowed by api.
It's a bit same as UART's current-speed... 
My advice would be to keep it, but make it clear in description this is  a default boot setting.

mnkp replied:
@erwango I understand that the your comment was actually related to "zephyr,speed" property. That's where I placed my reply.

mnkp replied:
The "zephyr,use-10-bit-addr" property really should go. Unless we want to add it and then remove shortly afterwards.

mnkp replied:
I've opened a Jira issue related to this problem: https://jira.zephyrproject.org/browse/ZEP-2373

ydamigos commented:
nxp,kinetis-i2c.yaml is missing

mnkp commented on dts/common/yaml/zephyr-i2c.yaml:
```diff
@@ -0,0 +1,19 @@
+    This binding gives Zephyr specific I2C properties
+
+properties:
+  - zephyr,speed :
```
I'm not convinced we need this property. The dts/common/yaml/i2c.yaml file defines "clock-frequency" property which defines maximum bus speed supported by all devices attached to the I2C bus. We can use it to set the default bus speed at boot. I can't think of a use case where we would like to set the bus speed at boot to a lower value.

mnkp replied:
In case we would like to keep both properties, one defining the suggested/default/current speed and one the maximum allowed speed we need to do it differently. The "clock-frequency" property defined in dts/common/yaml/i2c.yaml  is actually the suggested/default/current speed. That's the original meaning taken from Linux DTS which we, as I understand from the previous discussions, want to honor. We've changed the original description of "clock-frequency" from "Desired I2C bus clock frequency in Hz" to "Maximum I2C bus clock frequency in Hz" overloading its meaning. If we don't like the overloaded meaning we should return to its original one. And then create another property for the maximum allowed speed, if we deem it necessary. Otherwise it all becomes a bit messy.

mnkp commented on dts/arm/96b_carbon.fixup:
```diff
@@ -25,3 +25,4 @@
 #define CONFIG_I2C_1_NAME			ST_STM32_I2C_V1_40005400_LABEL
 #define CONFIG_I2C_1_EVENT_IRQ			ST_STM32_I2C_V1_40005400_EVENT
 #define CONFIG_I2C_1_ERROR_IRQ			ST_STM32_I2C_V1_40005400_ERROR
+#define CONFIG_I2C_1_DEFAULT_CFG		(ST_STM32_I2C_V1_40005400_ZEPHYR_SPEED << 1) | I2C_MODE_MASTER
```
The "| I2C_MODE_MASTER" in this place feels like a hack. I believe we need an independent DTS property to define master/slave mode at boot.

mnkp replied:
Defining Master Mode here is not clean. Even as a temporary solution until we have a proper master/slave support.

mnkp commented on drivers/i2c/Kconfig:
```diff
@@ -194,11 +195,11 @@ config I2C_0
 	string "Port 0 device name"
 	depends on I2C_0
-	default "I2C_0"
+	default "I2C_0" && !HAS_DTS_I2C
```
This was probably meant to be added to the line above.

galak commented:
Oops, thanks for catching that.

mnkp commented on dts/arm/hexiwear_k64.dts:
```diff
@@ -15,10 +16,12 @@
 		gpio_a = &gpioa;
-		gpio_b = &gpiob;
-		gpio_c = &gpioc;
+		gpio_b = &gpiob; gpio_c = &gpioc;
```
Was this change intended?

galak replied:
nope, weird

ydamigos commented:
I believe that we don't need to provide a default configuration for the I2C in DT, Kconfig or in the init function of the I2C controller.
The configuration of the I2C controller is tied, in most cases if not all, with the application/use case. So, the user application/test case/sample should handle the configuration of the I2C controller using the i2c_configure function from the I2C API, like the Zephyr's sensor drivers bmg160 and hp206c do.
That way, we remove some complexity from the DTS and Kconfig files. We could, probably, apply this to other drivers and remove more options from the DTS and Kconfig files, keeping them simpler.

MaureenHelm commented on drivers/i2c/i2c_mcux.c:
```diff
@@ -186,15 +186,15 @@ static const struct i2c_driver_api i2c_mcux_driver_api = {
 
 static const struct i2c_mcux_config i2c_mcux_config_0 = {
-	.base = I2C0,
+	.base = (void *)CONFIG_I2C_MCUX_0_BASE_ADDRESS,
```
Should cast to (I2C_Type *) instead of (void *). And if you're going to switch to the dts base address here, you should also do it for instance 1.

galak replied:
Thanks, missed the I2C1 case.  Will fix this up and split it out from the I2C DTS cfg patch for now.

MaureenHelm reviewed (changes requested):
This breaks the fxos8700 and fxas21002 sensor samples because they use I2C_0_NAME and I2C_1_NAME, which don't exist anymore.

galak commented:
Ok, need to see how I can handle I2C_0_NAME for the sensors

mnkp commented:
@ydamigos That's another valid approach. I wonder if there is an official Zephyr policy on how driver initialization should be handled?

That said, I'm not sure anymore if not initializing I2C driver at boot is such a good idea. I2C bus will be used by both user application and other Zephyr drivers. One such example is eth_sam_gmac Ethernet driver which reads the MAC value from an external I2C EEPROM during boot. After that it does not touch I2C bus ever again. If I2C driver is not configured at boot eth_sam_gmac driver should configure -> use -> disable it. What if there are more drivers which need to use the same I2C bus. Should every one of them configure -> use -> disable the I2C driver? There are also device drivers which need I2C bus to be enabled at all times possibly sharing the bus with the user application. One such device driver class are audio codecs which relay on I2C bus to send control messages to the codec chip. Managing all this is of course possible but complex.

Configuring I2C driver at boot and then treating it as a common resource any device driver or user application can use seems like a simpler and more robust approach.

ydamigos commented:
@mnkp You are right, we need to initialize the I2C driver at boot. I never thought about the use cases you mentioned above. I had in mind simpler applications.

galak commented:
Based on this we should probably be using clock-frequency to specify the initial bit rate of the bus.  I'll introduce these defines to be used with dts:

#define I2C_BITRATE_STANDARD	100000
#define I2C_BITRATE_FAST		400000
#define I2C_BITRATE_FAST_PLUS	1000000
#define I2C_BITRATE_HIGH		3400000
#define I2C_BITRATE_ULTRA		5000000

I'd rename the zephyr,use-10-bit-addr to zephyr,init-10-bit-addr

I'm leaving master/slave mode still alone w/regards to DT as we don't really support slave mode at all right now.  Still we do/or when we do we can address the DT props associated with slave mode.

mnkp commented:
As for the master/slave support I believe it would be easier and cleaner if we defined i2c_mode property in the .yaml file already and set it in the board .dts file as master. Then all the CONFIG_I2C_?_DEFAULT_CFG should be moved from .fixup to i2c.h. In the future when we add slave support there will be much less work, the DTS infrastructure will be all in place.

erwango commented:
Late comment, but I confirm that for dts based generation, it's required to have all default config/data
used at boot (such as .dev_config.raw = CONFIG_I2C_1_DEFAULT_CFG, in stm32 i2c driver)
populated in dts.

galak commented:
Replaced by #1081
